Providers/Profile.swift

```diff
@@ -443,27 +443,69 @@ open class BrowserProfile: Profile {
-                return
+        let command = SyncCommand.displayURIFromShareItem(item, asClient: id)
+        let fxaDeviceIds = clients.compactMap { $0.fxaDeviceId }
+
+        // If FxA Messages (Pushbox) is not enabled for this build, simply send the
+        // tabs using the old mechanism via Sync.
+        guard AppConstants.MOZ_FXA_MESSAGES else {
+            return self.remoteClientsAndTabs.insertCommands([command], forClients: clients) >>> {
+                self.syncManager.syncClients() >>> {
+                    account.notify(deviceIDs: fxaDeviceIds, collectionsChanged: ["clients"], reason: "sendtab")
+                    return succeed()
+                }
             }
-            
-            account.notify(deviceIDs: deviceIDs, collectionsChanged: ["clients"], reason: "sendtab")
         }
-        
-        return self.remoteClientsAndTabs.insertCommands(commands, forClients: clients) >>> {
-            let syncStatus = self.syncManager.syncClients()
-            syncStatus >>> notifyClients
-            return syncStatus
+
+        // TODO: If the RemoteDevice has the FxA Messages "availableCommands",
+        // send the tab via that instead.
+        // (e.g. https://hg.mozilla.org/mozilla-central/rev/8ef95c56df04#l2.16)
+        let deferredRemoteDevices = fxaDeviceIds.map { fxaDeviceId in
+            return self.remoteClientsAndTabs.getRemoteDevice(fxaDeviceId: fxaDeviceId)
+        }
+
+        let result = Success()
+
+        all(deferredRemoteDevices).upon { maybeRemoteDevices in
+            var newRemoteDevices: [FxADevice] = []
+            var oldRemoteClients: [RemoteClient] = []
+            let remoteDevices = maybeRemoteDevices.compactMap({ $0.successValue as? FxADevice })
+
+            for remoteDevice in remoteDevices {
+                if account.commandsClient.sendTab.isDeviceCompatible(remoteDevice) {
+                    newRemoteDevices.append(remoteDevice)
+                } else {
+                    if let oldRemoteClient = clients.find({ $0.fxaDeviceId == remoteDevice.id }) {
+                        oldRemoteClients.append(oldRemoteClient)
+                    }
+                }
+            }
+
+            if !newRemoteDevices.isEmpty {
+                // TODO: Check result and fallback to old API if necessary
+                account.commandsClient.sendTab.send(to: newRemoteDevices, url: item.url, title: item.title ?? "")
+            }
+
+            if oldRemoteClients.isEmpty {
+                result.fill(Maybe(success: ()))
+            } else {
+                self.remoteClientsAndTabs.insertCommands([command], forClients: oldRemoteClients) >>> {
+                    self.syncManager.syncClients() >>> {
+                        account.notify(deviceIDs: fxaDeviceIds, collectionsChanged: ["clients"], reason: "sendtab")
+                        result.fill(Maybe(success: ()))
+                    }
+                }
+            }
         }
+
+        return result
     }
```

Comment:
Most of this method is general to most commands. Can we pull it out such that the general bits only need writing once. e.g.

```swift
func emit(command: Command, to clients: [RemoteClient]) {
    let devices = …
    let capable = devices.filter { command.isDeviceCompatible($0) }
    let ids = Set(capable.map { $0.id })
    let incompatibleClients = clients.filter { !ids.contains($0.id) }

   command.emit(to: capable) >>== command.emitFallback(to: incompatibleClients)
}
```

This would be made much easier with `async`/`await`, of course.